haskell/test/SMT/Common/Gen.hs

```diff
@@ -0,0 +1,431 @@
+genRandomCalls = do
+  addrs <- get <&> gsAddresses <<&>> fst
+  sender1 <- Gen.element addrs
+
+  -- gen entrypoints
+  calls <- genProposingProcess
+  updateDelegateCall <- genUpdateDelegate <&> mkSimpleContractCall sender1
+  transferOwnershipCall <- genTransferOwnership <&> mkSimpleContractCall sender1
+  acceptOwnershipCall <- genAcceptOwnership <&> mkSimpleContractCall sender1
+  mkTransferContractCall <- genTransferContractTokensCall sender1
+
+  pure $ \args@ModelInputArg{..} ->
+    [ updateDelegateCall
+    , transferOwnershipCall
+    , acceptOwnershipCall
+    , mkTransferContractCall miaGovAddr
+    ] <> (calls args)
+
+  where
+    genTransferContractTokensCall :: Address -> GeneratorT (Address -> ModelCall)
+    genTransferContractTokensCall sender1 = do
+      f <- genTransferContractTokens
+      pure $ \govAddr -> mkSimpleContractCall sender1 $ f govAddr
+
+
+mkSimpleContractCall :: Address -> Parameter -> ModelCall
+mkSimpleContractCall sender1 param = do
+  ModelCall
+    { mcAdvanceLevel = Nothing
+    , mcParameter = param
+    , mcSource = ModelSource
+        { msoSource = sender1
+        , msoSender = sender1
+        }
+    }
+
+genProposingProcess :: GeneratorT (ModelInputArg -> [ModelCall])
+genProposingProcess = do
+  userAddrs <- get <&> gsAddresses
+  sender1 <- fst <$> Gen.element userAddrs
+  delegate1 <- fst <$> Gen.element userAddrs
+  invalidFrom <- fst <$> Gen.element userAddrs
+
+  mkGenPropose <- get <&> gsMkGenPropose
+  mkPropose <- mkGenPropose sender1 delegate1 invalidFrom
+
+  mkGenCustomCalls <- get <&> gsMkCustomCalls
+  mkCustomCalls <- mkGenCustomCalls
+
+  -- gen entrypoints
+  flushAction <- genFlush
+  mkVote <- genVote
+  mkDropProposal <- genDropProposal
+
+  -- | `advanceLvl` is provided precisely to ensure the order of operations results in successful proposal.
+  let mkCall param advanceLvl = ModelCall
+        { mcAdvanceLevel = advanceLvl
+        , mcParameter = param
+        , mcSource = ModelSource
+          { msoSource = sender1
+          , msoSender = sender1
+          }
+        }
+
+  pure $ \args@ModelInputArg{..} ->
+    let
+      (proposeAction, metaSize, key) = mkPropose miaGuardianAddr miaGovAddr
+      freezeAmt = metaSize + fixedFee
+      permitProtecteds = mkVote key freezeAmt
+      voterFreezeAmt = getVoteFreezeAmt permitProtecteds
+      dropProposalAction = mkDropProposal key
+    in
+      [ mkCall (XtzForbidden $ Update_delegate [DelegateParam True delegate1]) Nothing
+      , mkCall (XtzForbidden $ Freeze (#amount .! freezeAmt) ) Nothing
+      , mkCall (XtzForbidden $ Freeze (#amount .! voterFreezeAmt) ) Nothing
+      , mkCall proposeAction (Just 1)
+      , mkCall (XtzForbidden $ Vote permitProtecteds) (Just 1)
+      , mkCall flushAction (Just 2)
+      , mkCall (XtzForbidden $ Unfreeze (#amount .! freezeAmt) ) Nothing
+      , mkCall dropProposalAction Nothing
+      ] <> ((\param -> mkCall (XtzAllowed $ CallCustom param) Nothing) <$> mkCustomCalls args)
```

Comment:
Am I reading this right: do we always create a proposal, vote on it, flush it and drop it, in this order?

Reply:
Yeah, we call them in order to make them more likely to result in successful case.

Reply:
This seems to leave out a lot of interesting scenarios though, a few off the top of my head:

1. multiple votes on the same proposal
2. "tie" scenario
2. no votes on a proposal
2. dropping a proposal that hasn't been voted on
3. dropping a proposal that hasn't been flushed
4. flushing a proposal twice
5. dropping a proposal twice

Reply:
There is another `genRandomCalls` which is intended to generate all these calls and some other in random order (Currently, it is not in random order. I'll fix this)

With this, it can hits some of the scenarios above if we run it long enough. Still, it won't hit all those cases, I'm thinking of handling this in a separate issue if you think it's alright.

Reply:
Yes of course, let's create another issue, handle this separately, and take into account the points discussed in the last slack conversation.

Reply:
I've updated `genRandomCalls` to call entrypoints in random order. This would help it to be less rigid.

Also added an issue to improve the generator: https://github.com/tqtezos/baseDAO/issues/295